src/js/serial_backend.js

```diff
@@ -782,53 +784,45 @@ function startLiveDataRefreshTimer() {
-
-    gui_log(i18n.getMessage('deviceRebooting'));
-
-    // Close connection gracefully if it still exists.
     const previousTimeStamp = connectionTimestamp;
-
-    if (serial.connectionId) {
-        if (GUI.connected_to || GUI.connecting_to) {
-            $('a.connect').trigger('click');
-        } else {
-            serial.disconnect();
-        }
-    }
+    gui_log(i18n.getMessage('deviceRebooting'));
 
     // In virtual mode reconnect when autoconnect is enabled
-    if (isVirtual) {
-        return setTimeout(() => {
-            if (PortHandler.portPicker.autoConnect) {
-                $('a.connect').trigger('click');
-            }
-            if (typeof callback === 'function') {
-                callback();
-            }
+    if (PortHandler.portPicker.selectedPort === 'virtual' && PortHandler.portPicker.autoConnect) {
+        return setTimeout(function() {
+            $('a.connect').trigger('click');
         }, 500);
     }
 
-    // Wait for serial or tcp connection to be available
-    let attempts = 0;
-    const reconnect = setInterval(waitforSerial, 100);
+    function validateConnection() {
+        if (PortHandler.portPicker.selectedPort && PortHandler.portPicker.autoConnect) {
+            $('a.connect').trigger('click');
+        }
 
-    function waitforSerial() {
-        if ((connectionTimestamp !== previousTimeStamp && CONFIGURATOR.connectionValid) || GUI.active_tab === 'firmware_flasher') {
-            console.log(`Serial connection available after ${attempts / 10} seconds`);
-            clearInterval(reconnect);
-            gui_log(i18n.getMessage('deviceReady'));
+        // Wait for serial or tcp connection to be available
+        let attempts = 0;
+        const reconnect = setInterval(waitforSerial, 100);
 
-            if (typeof callback === 'function') {
-                callback();
-            }
-        } else {
-            attempts++;
-            if (attempts > 100) {
+        function waitforSerial() {
+            if ((connectionTimestamp !== previousTimeStamp && CONFIGURATOR.connectionValid) || GUI.active_tab === 'firmware_flasher') {
+                console.log(`Serial connection available after ${(attempts / 10) + 2} seconds`);
                 clearInterval(reconnect);
-                console.log(`failed to get serial connection, gave up after 10 seconds`);
-                gui_log(i18n.getMessage('serialPortOpenFail'));
+                gui_log(i18n.getMessage('deviceReady'));
+                if (typeof callback === 'function') {
+                    callback();
+                }
+            } else {
+                attempts++;
+                if (attempts > 100) {
+                    clearInterval(reconnect);
+                    console.log(`failed to get serial connection, gave up after 10 seconds`);
+                    gui_log(i18n.getMessage('serialPortOpenFail'));
+                }
             }
         }
     }
```

Comment:
Again, after my PR refactoring: https://github.com/betaflight/betaflight-configurator/pull/3984 there is an easier way to do this: simply put the timestamp into a variable, `rebootTimestamp`, and after receiving the event of new device if this timestamp is small, connect. Easier, without loops, and without polling. I can modify it after both are merged.

Reply:
Agree, now I moved MSP_SET_REBOOT to reinitialize function timestamp would work better as in the callback it already was updated before comparison.